// @applitools/eyes-api/dist/output/RenderingInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderingInfoData = void 0;
const utils = __importStar(require("@applitools/utils"));
class RenderingInfoData {
    /** @internal */
    constructor(info) {
        this._info = {};
        this._info = info instanceof RenderingInfoData ? info.toJSON() : info;
    }
    get accessToken() {
        return this._info.accessToken;
    }
    getAccessToken() {
        return this.accessToken;
    }
    getDecodedAccessToken() {
        return utils.general.jwtDecode(this._info.accessToken);
    }
    /** @deprecated */
    setAccessToken(accessToken) {
        this._info.accessToken = accessToken;
    }
    get serviceUrl() {
        return this._info.serviceUrl;
    }
    getServiceUrl() {
        return this.serviceUrl;
    }
    /** @deprecated */
    setServiceUrl(serviceUrl) {
        this._info.serviceUrl = serviceUrl;
    }
    get resultsUrl() {
        return this._info.resultsUrl;
    }
    getResultsUrl() {
        return this.resultsUrl;
    }
    /** @deprecated */
    setResultsUrl(resultsUrl) {
        this._info.resultsUrl = resultsUrl;
    }
    get stitchingServiceUrl() {
        return this._info.stitchingServiceUrl;
    }
    getStitchingServiceUrl() {
        return this.stitchingServiceUrl;
    }
    /** @deprecated */
    setStitchingServiceUrl(stitchingServiceUrl) {
        this._info.stitchingServiceUrl = stitchingServiceUrl;
    }
    /** @internal */
    toObject() {
        return this._info;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._info);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.RenderingInfoData = RenderingInfoData;
