// @applitools/eyes-api/dist/output/RunningSession.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunningSessionData = void 0;
const utils = __importStar(require("@applitools/utils"));
const RenderingInfo_1 = require("./RenderingInfo");
class RunningSessionData {
    /** @internal */
    constructor(session) {
        this._session = {};
        this._session = session instanceof RunningSessionData ? session.toJSON() : session;
    }
    get id() {
        return this._session.id;
    }
    getId() {
        return this.id;
    }
    setId(id) {
        this._session.id = id;
    }
    get sessionId() {
        return this._session.sessionId;
    }
    getSessionId() {
        return this.sessionId;
    }
    /** @deprecated */
    setSessionId(sessionId) {
        this._session.sessionId = sessionId;
    }
    get batchId() {
        return this._session.batchId;
    }
    getBatchId() {
        return this.batchId;
    }
    /** @deprecated */
    setBatchId(batchId) {
        this._session.batchId = batchId;
    }
    get baselineId() {
        return this._session.baselineId;
    }
    getBaselineId() {
        return this.baselineId;
    }
    /** @deprecated */
    setBaselineId(baselineId) {
        this._session.baselineId = baselineId;
    }
    get url() {
        return this._session.url;
    }
    getUrl() {
        return this.url;
    }
    /** @deprecated */
    setUrl(url) {
        this._session.url = url;
    }
    get isNew() {
        return this._session.isNew;
    }
    getIsNew() {
        return this.isNew;
    }
    /** @deprecated */
    setIsNew(isNew) {
        this._session.isNew = isNew;
    }
    get renderingInfo() {
        return this._session.renderingInfo;
    }
    getRenderingInfo() {
        return new RenderingInfo_1.RenderingInfoData(this.renderingInfo);
    }
    /** @deprecated */
    setRenderingInfo(renderingInfo) {
        this._session.renderingInfo = renderingInfo;
    }
    /** @internal */
    toObject() {
        return this._session;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._session);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.RunningSessionData = RunningSessionData;
