// @applitools/eyes-api/dist/output/TestResults.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestResultsData = void 0;
const utils = __importStar(require("@applitools/utils"));
const TestResultsStatus_1 = require("../enums/TestResultsStatus");
const RectangleSize_1 = require("../input/RectangleSize");
const SessionUrls_1 = require("./SessionUrls");
const StepInfo_1 = require("./StepInfo");
class TestResultsData {
    /** @internal */
    constructor(results, deleteTest) {
        this._results = {};
        this._deleteTest = deleteTest;
        if (!results)
            return this;
        this._results = results instanceof TestResultsData ? results.toJSON() : results;
    }
    get id() {
        return this._results.id;
    }
    getId() {
        return this.id;
    }
    /** @deprecated */
    setId(id) {
        this._results.id = id;
    }
    get name() {
        return this._results.name;
    }
    getName() {
        return this.name;
    }
    /** @deprecated */
    setName(name) {
        this._results.name = name;
    }
    get secretToken() {
        return this._results.secretToken;
    }
    getSecretToken() {
        return this.secretToken;
    }
    /** @deprecated */
    setSecretToken(secretToken) {
        this._results.secretToken = secretToken;
    }
    get status() {
        return this._results.status;
    }
    getStatus() {
        return this.status;
    }
    /** @deprecated */
    setStatus(status) {
        this._results.status = status;
    }
    get appName() {
        return this._results.appName;
    }
    getAppName() {
        return this.appName;
    }
    /** @deprecated */
    setAppName(appName) {
        this._results.appName = appName;
    }
    get batchName() {
        return this._results.batchName;
    }
    getBatchName() {
        return this.batchName;
    }
    /** @deprecated */
    setBatchName(batchName) {
        this._results.batchName = batchName;
    }
    get batchId() {
        return this._results.batchId;
    }
    getBatchId() {
        return this.batchId;
    }
    /** @deprecated */
    setBatchId(batchId) {
        this._results.batchId = batchId;
    }
    get branchName() {
        return this._results.batchName;
    }
    getBranchName() {
        return this.branchName;
    }
    /** @deprecated */
    setBranchName(branchName) {
        this._results.branchName = branchName;
    }
    get hostOS() {
        return this._results.hostOS;
    }
    getHostOS() {
        return this.hostOS;
    }
    /** @deprecated */
    setHostOS(hostOS) {
        this._results.hostOS = hostOS;
    }
    get hostApp() {
        return this._results.hostApp;
    }
    getHostApp() {
        return this.hostApp;
    }
    /** @deprecated */
    setHostApp(hostApp) {
        this._results.hostApp = hostApp;
    }
    get hostDisplaySize() {
        return this._results.hostDisplaySize;
    }
    getHostDisplaySize() {
        return new RectangleSize_1.RectangleSizeData(this.hostDisplaySize);
    }
    /** @deprecated */
    setHostDisplaySize(hostDisplaySize) {
        this._results.hostDisplaySize = hostDisplaySize;
    }
    get accessibilityStatus() {
        return this._results.accessibilityStatus;
    }
    getAccessibilityStatus() {
        return this.accessibilityStatus;
    }
    /** @deprecated */
    setAccessibilityStatus(accessibilityStatus) {
        this._results.accessibilityStatus = accessibilityStatus;
    }
    get startedAt() {
        return this._results.startedAt;
    }
    getStartedAt() {
        return new Date(this.startedAt);
    }
    /** @deprecated */
    setStartedAt(startedAt) {
        this._results.startedAt = startedAt;
    }
    get duration() {
        return this._results.duration;
    }
    getDuration() {
        return this.duration;
    }
    /** @deprecated */
    setDuration(duration) {
        this._results.duration = duration;
    }
    get isNew() {
        return this._results.isNew;
    }
    getIsNew() {
        return this.isNew;
    }
    /** @deprecated */
    setIsNew(isNew) {
        this._results.isNew = isNew;
    }
    get isDifferent() {
        return this._results.isDifferent;
    }
    getIsDifferent() {
        return this.isDifferent;
    }
    /** @deprecated */
    setIsDifferent(isDifferent) {
        this._results.isDifferent = isDifferent;
    }
    get isAborted() {
        return this._results.isAborted;
    }
    getIsAborted() {
        return this.isAborted;
    }
    /** @deprecated */
    setIsAborted(isAborted) {
        this._results.isAborted = isAborted;
    }
    get appUrls() {
        return this._results.appUrls;
    }
    getAppUrls() {
        return new SessionUrls_1.SessionUrlsData(this.appUrls);
    }
    /** @deprecated */
    setAppUrls(appUrls) {
        this._results.appUrls = appUrls;
    }
    get apiUrls() {
        return this._results.apiUrls;
    }
    getApiUrls() {
        return new SessionUrls_1.SessionUrlsData(this.apiUrls);
    }
    /** @deprecated */
    setApiUrls(apiUrls) {
        this._results.apiUrls = apiUrls;
    }
    get stepsInfo() {
        return this._results.stepsInfo;
    }
    getStepsInfo() {
        return this.stepsInfo.map(info => new StepInfo_1.StepInfoData(info));
    }
    /** @deprecated */
    setStepsInfo(stepInfo) {
        this._results.stepsInfo = stepInfo;
    }
    get steps() {
        return this._results.steps;
    }
    getSteps() {
        return this.steps;
    }
    /** @deprecated */
    setSteps(steps) {
        this._results.steps = steps;
    }
    get matches() {
        return this._results.matches;
    }
    getMatches() {
        return this.matches;
    }
    /** @deprecated */
    setMatches(matches) {
        this._results.matches = matches;
    }
    get mismatches() {
        return this._results.mismatches;
    }
    getMismatches() {
        return this.mismatches;
    }
    /** @deprecated */
    setMismatches(mismatches) {
        this._results.mismatches = mismatches;
    }
    get missing() {
        return this._results.missing;
    }
    getMissing() {
        return this.missing;
    }
    /** @deprecated */
    setMissing(missing) {
        this._results.missing = missing;
    }
    get exactMatches() {
        return this._results.exactMatches;
    }
    getExactMatches() {
        return this.exactMatches;
    }
    /** @deprecated */
    setExactMatches(exactMatches) {
        this._results.exactMatches = exactMatches;
    }
    get strictMatches() {
        return this._results.strictMatches;
    }
    getStrictMatches() {
        return this.strictMatches;
    }
    /** @deprecated */
    setStrictMatches(strictMatches) {
        this._results.strictMatches = strictMatches;
    }
    get contentMatches() {
        return this._results.contentMatches;
    }
    getContentMatches() {
        return this.contentMatches;
    }
    /** @deprecated */
    setContentMatches(contentMatches) {
        this._results.contentMatches = contentMatches;
    }
    get layoutMatches() {
        return this._results.layoutMatches;
    }
    getLayoutMatches() {
        return this.layoutMatches;
    }
    /** @deprecated */
    setLayoutMatches(layoutMatches) {
        this._results.layoutMatches = layoutMatches;
    }
    get noneMatches() {
        return this._results.noneMatches;
    }
    getNoneMatches() {
        return this.noneMatches;
    }
    /** @deprecated */
    setNoneMatches(noneMatches) {
        this._results.noneMatches = noneMatches;
    }
    get url() {
        return this._results.url;
    }
    getUrl() {
        return this.url;
    }
    /** @deprecated */
    setUrl(url) {
        this._results.url = url;
    }
    isPassed() {
        return this._results.status === TestResultsStatus_1.TestResultsStatusEnum.Passed;
    }
    async delete() {
        if (!this._deleteTest)
            return;
        return this._deleteTest({ testId: this.id, batchId: this.batchId, secretToken: this.secretToken });
    }
    /** @deprecated */
    async deleteSession() {
        await this.delete();
    }
    /** @internal */
    toObject() {
        return this._results;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._results);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.TestResultsData = TestResultsData;
